lambda/matching.ml

```diff
@@ -690,6 +683,70 @@ end = struct
       | _ ->
           pretty_matrix Format.err_formatter pss;
           fatal_error "Matching.Default_environment.specialize_matrix"
+
+    (* Filter just one row, without a `rem` accumulator
+       of further rows to process.
+       The following equality holds:
+         filter_rec ((p :: ps) :: rem)
+         = filter_one p ps @ filter_rec rem
+    *)
+    and filter_one p ps =
+      filter_rec [ p :: ps ]
+
+    and filter_rec_or p1 p2 ps rem =
+      match arity with
+      | 0 -> (
+          (* if K has arity 0, specializing ((K|K)::rem) returns just (rem):
+             if either sides works (filters into a non-empty list),
+             no need to keep the other. *)
+          match filter_one p1 ps with
+          | [] -> filter_rec ((p2 :: ps) :: rem)
+          | matches -> matches @ filter_rec rem
+        )
+      | 1 -> (
+          (* if K has arity 1, ((K p | K q) :: rem) can be expressed
+             as ((p | q) :: rem): even if both sides of an or-pattern
+             match, we can compress the output in a single row,
+             instead of duplicating the row.
+
+             In particular, we respect the invariant that filtering
+             a single row (the filter_one calls) produces either an
+             empty list or a single-row output list.
+
+             Given that we are in the arity-one case, we know that the
+             ouput row (if any) is non-empty, with exactly one new
+             pattern consed to the tail [ps] of our input row.
+          *)
+          match (filter_one p1 ps, filter_one p2 ps) with
+          | [], row
+          | row, [] ->
+              row @ filter_rec rem
+          | [ (arg1 :: _) ], [ (arg2 :: _) ] ->
+              (* the wildcard patterns can only be ps *)
+              (* the output for the row (p1|p2)::ps is of size 1,
```

Comment:
The comment seems a bit unclear, it might be clearer to state that the row `(p1|p2) :: ps` has size `1 + List.length ps` respecting the invariant.

Reply:
I meant that there is a single row in the output, so not the invariant you are thinking of (but the one that is described as "the invariant" in the comment before). I have to clarify this.

Reply:
I tried to clarify those comments, let me know what you think.

Reply:
Thanks,  giving names to the invariants is clearer.